src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -1359,6 +1539,22 @@ def get_connected_passed_kwargs(prefix):
         model.register_to_config(_name_or_path=pretrained_model_name_or_path)
         return model
 
+    def _assign_components_to_devices(sorted_components: dict, devices: dict):
+        device_ids = list(devices.keys())
+        device_cycle = device_ids + device_ids[::-1]
+
+        mapping = {}
+        current_device_index = 0
+        for component in sorted_components:
+            device_id = device_cycle[current_device_index % len(device_cycle)]
+            if device_id not in mapping:
+                mapping[device_id] = [component]
+            else:
+                mapping[device_id].append(component)
+            current_device_index += 1
+
+        return mapping
+
```

Comment:
TODO for later: use the memory req for each module and max_memory of each gpus.

Reply:
Could you elaborate this a bit? Currently, we follow the strategy from: 
https://github.com/huggingface/diffusers/pull/6396#issuecomment-1918988743.

Are you proposing the following?

Place a given module based on its memory requirement and the available memory of a given device? This could probably be a different strategy, WDYT?

Reply:
I mean we should at least check that the module will fit the given device when we are doing this assignment, and if that's not the case, I think we should probably either return an error or decide to offload the module to the cpu.

Reply:
Do you think we should do it in this PR? I think we should. WDYT?

Reply:
yes that would be better !

Reply:
LMK if the changes in 1cd6271b are what you're envisioning.

Reply:
yes, that's better !